apis/v1beta1/config.go

```diff
@@ -445,64 +441,46 @@ const (
 // because the port is used to generate Service objects and mappings.
 func (s *Service) MetricsEndpoint(logger logr.Logger) (string, int32, error) {
 	telemetry := s.GetTelemetry()
-	if telemetry == nil || telemetry.Metrics.Address == "" {
+	if telemetry == nil {
 		return defaultServiceHost, defaultServicePort, nil
 	}
 
-	// The regex below matches on strings that end with a colon followed by the environment variable expansion syntax.
-	// So it should match on strings ending with: ":${env:POD_IP}" or ":${POD_IP}".
-	const portEnvVarRegex = `:\${[env:]?.*}$`
-	isPortEnvVar := regexp.MustCompile(portEnvVarRegex).MatchString(telemetry.Metrics.Address)
-	if isPortEnvVar {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg)
-		return "", 0, errors.New(errMsg)
-	}
-
-	// The regex below matches on strings that end with a colon followed by 1 or more numbers (representing the port).
-	const explicitPortRegex = `:(\d+$)`
-	explicitPortMatches := regexp.MustCompile(explicitPortRegex).FindStringSubmatch(telemetry.Metrics.Address)
-	if len(explicitPortMatches) <= 1 {
-		return telemetry.Metrics.Address, defaultServicePort, nil
-	}
+	if telemetry.Metrics.Address != "" && len(telemetry.Metrics.Readers) == 0 {
+		host, port, err := parseAddressEndpoint(telemetry.Metrics.Address)
+		if err != nil {
+			return "", 0, err
+		}
 
-	port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
-	if err != nil {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg, "error", err)
-		return "", 0, err
+		return host, port, nil
 	}
 
-	host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
-	return host, int32(port), nil
+	return defaultServiceHost, defaultServicePort, nil
 }
 
 // ApplyDefaults inserts configuration defaults if it has not been set.
 func (s *Service) ApplyDefaults(logger logr.Logger) error {
-	telemetryAddr, telemetryPort, err := s.MetricsEndpoint(logger)
+	tel := s.GetTelemetry()
+
+	if tel == nil {
+		tel = &Telemetry{}
+	}
+
+	host, port, err := s.MetricsEndpoint(logger)
 	if err != nil {
 		return err
 	}
 
-	tm := &AnyConfig{
-		Object: map[string]interface{}{
-			"metrics": map[string]interface{}{
-				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
-			},
-		},
+	if tel.Metrics.Address != "" {
+		tel.Metrics.Address = fmt.Sprintf("%s:%d", host, port)
```

Comment:
Is there a reason to change the address here?

Reply:
Yes.
1. If you check the code before this PR, `ApplyDefaults` is setting a correct value for `address`. 
2. In this PR, I was removing the value from `address` to ensure that when we set a reader, there is not an issue due to two Prometheus servers listening on the same port.
3. I was requested to remove point 2.
4. I reverted to set a proper value for `address` when the user sets `address` and not add any readers. Which makes sense because there can be users that are still creating `OpenTelemetryCollector` instances using the deprecated field. Some of our tests are doing that.
5. If the field `address` was set in an `OpenTelemetryCollector` instance running in a previous version of the operator, the upgrade mechanism will remove the usage of `address`.

Reply:
But if address is already set, why do we need to modify it? My naive view of this logic should be something like "If neither `address` nor `readers` are set, add a default reader. Otherwise, do nothing". Is there an issue with that?

Reply:
Actually not. Sorry this comes from the mindset of the previous solution.

Reply:
> But if address is already set, why do we need to modify it? My naive view of this logic should be something like "If neither `address` nor `readers` are set, add a default reader. Otherwise, do nothing". Is there an issue with that?

I posted a comment, removed, replied something different and later you posted again. 
![Screenshot 2025-04-09 at 13 37 08](https://github.com/user-attachments/assets/72302361-59d5-4cef-a654-62150640ae91)

I fixed it.